src/auth/models.cpp

```diff
@@ -183,28 +202,38 @@ bool operator==(const Permissions &first, const Permissions &second) {
 
 bool operator!=(const Permissions &first, const Permissions &second) { return !(first == second); }
 
-const std::string ASTERISK = "*";
-
-FineGrainedAccessPermissions::FineGrainedAccessPermissions(const std::unordered_set<std::string> &grants,
-                                                           const std::unordered_set<std::string> &denies)
+FineGrainedAccessPermissions::FineGrainedAccessPermissions(const std::unordered_map<std::string, uint64_t> &grants,
+                                                           const std::unordered_map<std::string, uint64_t> &denies)
     : grants_(grants), denies_(denies) {}
 
-PermissionLevel FineGrainedAccessPermissions::Has(const std::string &permission) const {
-  if ((denies_.size() == 1 && denies_.find(ASTERISK) != denies_.end()) || denies_.find(permission) != denies_.end()) {
+PermissionLevel FineGrainedAccessPermissions::Has(const std::string &permission, LabelPermission label_permission) {
+  if (denies_.size() == 1 && denies_.find(ASTERISK) != denies_.end()) {
     return PermissionLevel::DENY;
   }
 
-  if ((grants_.size() == 1 && grants_.find(ASTERISK) != grants_.end()) || grants_.find(permission) != denies_.end()) {
+  if ((grants_.size() == 1 && grants_.find(ASTERISK) != grants_.end())) {
     return PermissionLevel::GRANT;
   }
 
-  return PermissionLevel::NEUTRAL;
+  auto grants_permission = PermissionLevel::DENY;
+  auto denies_permission = PermissionLevel::GRANT;
```

Comment:
why is this reversed?

Reply:
Those are just initail values. Logically, if something is not in `denies_` it is granted, likewise if something is not in `grants_` it is denied